staging/src/k8s.io/client-go/tools/cache/reflector.go

```diff
@@ -210,6 +214,17 @@ func (r *Reflector) ListAndWatch(stopCh <-chan struct{}) error {
 			}))
 			if r.WatchListPageSize != 0 {
 				pager.PageSize = r.WatchListPageSize
+			} else {
+				// User didin't explicitly requested pagination.
+				if options.ResourceVersion != "" {
+					// With ResourceVersion != "", we have a possibility to list from watch cache,
+					// but we do that (for ResourceVersion != "0") only if Limit is unset.
+					// To avoid thundering herd on etcd (e.g. on master upgrades), we explicitly
+					// switch off pagination to force listing from watch cache (if enabled).
```

Comment:
do we know what this does on high-cardinality resources without the watch cache (like events)? it would be good to know the following things:
* what possible errors does a client see if the etcd server is overloaded (timeout? internal server error? 429? something else?)
* does the client ever change anything in response to errors it gets back from an unpaged-with-resourceVersion request (either adding paging back in, or dropping back to a consistent read request, or something else) or does it just keep looping on this request that might kill a high-cardinality uncached resource?

Reply:
It's unfortunately not 429.
Those would be 5XX errors - timeouts, gateway timeouts, etc.

> does the client ever change anything in response to errors it gets back from an unpaged-with-resourceVersion request (either adding paging back in, or dropping back to a consistent read request, or something else) or does it just keep looping on this request that might kill a high-cardinality uncached resource?

To clarify this one - do you suggest that we fallback to RV="" requests in case of some specific errors (like the ones that I mentioned above)?
[Currently nothing like that happens, and clients don't have any way to fix it (other then start setting page size explicitly in reflector.]

Reply:
currently clients request `resourceVersion=0&limit=10000`, so resources without the watch cache enabled are paged, right?

Reply:
> Currently clients request resourceVersion=0&limit=10000, so resources without the watch cache enabled are paged, right?

&limit=500, but yes - without watchcache those are paged.

Reply:
RV="0" is a special case where watch cache is ignoring the limit anyway - so I reverted this case so not regress in this case.
It doesn't fully solve the problem, but it at least solves the initialization that happens every single time.

Reply:
to summarize the scenario where we are still exposed:
1. informers list/watching high-cardinality resources that do not have the watch cache enabled
    * example resources: events (by default), any resource (by config, e.g. clusters that disable watch cache on secrets)
    * example clients: kube-controller-manager (for quota and namespace controllers)
2. client's last-observed version is too old (compacted or out of watch window) because resource is low-velocity or client communications with the apiserver were interrupted
3. re-watch fails with "version too old"
4. client re-lists with specific resourceVersion, this PR omits paging (hoping we'll hit the watch cache)
5. full list from etcd times out, client retries... goto 4

What will detect/interrupt the 4-5-4-5-4-5-4-5-... cycle?